x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/api.ts

```diff
@@ -33,37 +36,101 @@ import {
 } from './types';
 import { KibanaServices } from '../../../../common/lib/kibana';
 import * as i18n from '../../../pages/detection_engine/rules/translations';
+import { RulesSchema, rulesSchema } from '../../../../../common/detection_engine/schemas/response';
+import { validateEither } from '../../../../../common';
+import {
+  createRulesSchema,
+  updateRulesSchema,
+  patchRulesSchema,
+} from '../../../../../common/detection_engine/schemas/request';
+import { toError, toPromise } from '../../../../../common/shared_imports';
+
+/**
+ * Create provided Rule
+ *
+ * @param rule CreateRulesSchema to add
+ * @param signal to cancel request
+ *
+ * @throws An error if response is not OK
+ */
+const createRule = async ({ rule, signal }: CreateRulesProps): Promise<RulesSchema> =>
+  KibanaServices.get().http.fetch<RulesSchema>(DETECTION_ENGINE_RULES_URL, {
+    method: 'POST',
+    body: JSON.stringify(rule),
+    signal,
+  });
+
+const createRuleWithValidation = async ({ rule, signal }: CreateRulesProps): Promise<RulesSchema> =>
+  pipe(
+    rule,
+    (body) => fromEither(validateEither(createRulesSchema, body)),
+    chain((payload) => tryCatch(() => createRule({ signal, rule: { ...payload } }), toError)),
+    chain((response) => fromEither(validateEither(rulesSchema, response))),
+    flow(toPromise)
+  );
+
+export { createRuleWithValidation as createRule };
 
 /**
- * Add provided Rule
+ * Update provided Rule using PUT
  *
- * @param rule to add
+ * @param rule UpdateRulesSchema to be updated
  * @param signal to cancel request
  *
  * @throws An error if response is not OK
  */
-export const addRule = async ({ rule, signal }: AddRulesProps): Promise<NewRule> =>
-  KibanaServices.get().http.fetch<NewRule>(DETECTION_ENGINE_RULES_URL, {
-    method: rule.id != null ? 'PUT' : 'POST',
+const updateRule = async ({ rule, signal }: UpdateRulesProps): Promise<RulesSchema> =>
+  KibanaServices.get().http.fetch<RulesSchema>(DETECTION_ENGINE_RULES_URL, {
+    method: 'PUT',
     body: JSON.stringify(rule),
     signal,
   });
 
+const updateRuleWithValidation = async ({ rule, signal }: UpdateRulesProps): Promise<RulesSchema> =>
+  pipe(
+    rule,
+    (body) => fromEither(validateEither(updateRulesSchema, body)),
+    chain((payload) => tryCatch(() => updateRule({ signal, rule: { ...payload } }), toError)),
+    chain((response) => fromEither(validateEither(rulesSchema, response))),
+    flow(toPromise)
+  );
+
+export { updateRuleWithValidation as updateRule };
+
 /**
- * Patch provided Rule
+ * Patch provided rule
```

Comment:
If we're going to discourage usage of patch in a certain way, we should probably limit the schema so that we can only patch the fields that we're expecting to be patched.

Reply:
That's a good call. I think it's particular to the way we built the UI (need to double check that) so maybe creating a limited schema in the front end makes sense.

Reply:
If I recall correctly, usage of patch was highly discouraged in general... so maybe worth discussing limiting on the backend too? Shouldn't block this PR, but if we could prevent someone from trying to use the API in ways that could be detrimental, I think that could be worthwhile.

Reply:
Sorry, just circling back on this PR now. I think that's definitely a good point that we should discuss. I added it to our team discussion for next week.